dlib/cuda/cudnn_dlibapi.cpp

```diff
@@ -1021,36 +1056,18 @@ namespace dlib
                 tensor_descriptor dest_desc;
                 dest_desc.set_size(out_num_samples,out_k,out_nr,out_nc);
 
-                select_best_algorithms(data, dest_desc);
-
-                CHECK_CUDNN(cudnnGetConvolutionForwardWorkspaceSize( 
-                        context(),
-                        descriptor(data),
-                        (const cudnnFilterDescriptor_t)filter_handle,
-                        (const cudnnConvolutionDescriptor_t)conv_handle,
-                        descriptor(dest_desc),
-                        (cudnnConvolutionFwdAlgo_t)forward_algo,
-                        &forward_workspace_size_in_bytes));
-
-
-                CHECK_CUDNN(cudnnGetConvolutionBackwardDataWorkspaceSize(
-                        context(),
-                        (const cudnnFilterDescriptor_t)filter_handle,
-                        descriptor(dest_desc),
-                        (const cudnnConvolutionDescriptor_t)conv_handle,
-                        descriptor(data),
-                        (cudnnConvolutionBwdDataAlgo_t)backward_data_algo,
-                        &backward_data_workspace_size_in_bytes));
-
-
-                CHECK_CUDNN(cudnnGetConvolutionBackwardFilterWorkspaceSize( 
-                        context(),
-                        descriptor(data),
-                        descriptor(dest_desc),
-                        (const cudnnConvolutionDescriptor_t)conv_handle,
-                        (const cudnnFilterDescriptor_t)filter_handle,
-                        (cudnnConvolutionBwdFilterAlgo_t)backward_filters_algo,
-                        &backward_filters_workspace_size_in_bytes));
+                try
+                {
+                    select_best_algorithms(data, dest_desc, allow_cache_use::yes);
+                    update_convolution_data_workspace_sizes(data, dest_desc);
+                }
+                catch (dlib::cudnn_error&)
+                {
+                    // Sometimes the values stored in `config_to_algo_cache` do not quite work -
+                    // so let's get a fresh estimate, instead of using a cached value.
+                    select_best_algorithms(data, dest_desc, allow_cache_use::no);
+                    update_convolution_data_workspace_sizes(data, dest_desc);
```

Comment:
Could have made it so that we don't try to select the best algorithms again, if the first estimate did not come from the cache. But wanted to keep this simple.